### PR TITLE
Add optional pushStartingAt threshold parameter to update event array constructor

### PR DESCRIPTION
**Pull request by daehn**
# What's in this PR?

When parsing the notification stream in response to a push notification we want to mark all notifications that occurred after a given UUID (type 1) as push event source.
To achieve this the `ZMUpdateEvent` array factory method now has an optional `pushStartingAfter` parameter, if provided, all events greater than or equal to the given uuid will be marked as `ZMUpdateEventSourcePushNotification` while older ones will remain `ZMUpdateEventSourceWebSocket`. The reasoning behind this is that we want treat these update events as if they would have been received through a push notification in order to generate a local push.

With this we also have the option to pass in a computed UUID that is a couple of minutes in the past. That would allow missed notifications to be fully displayed, e.g. if the user was under bad network and received a fallback, he might receive another message 5 min afterwards when his network conditions recovered. If we pass in a uuid with a timestamp 10 min in the past we would display the full message text. The downside of this would of course be that the user would have one fallback and the full text notification for this message.

### CODE REVIEW

**Comment by daehn** [comment]
The notification dispatcher consumes all events like a transcoder, and if the source is `.pushNotification` and a couple of other requirements are met (like the notification not having been displayed yet) a notification will be shown. As we are now fetching the notification stream all events will be of source type `.webSocket` and will not create a notification by default. The missing update events transcoder then checks if we are in the background and if the fetch has been triggered by an incoming push notification, but then we still only want to set this flag for events with a uuid newer or equal to the notification uuid that triggered the fetch initially. 
So basically we do not want to create pushes for all events if we are in the background (there is another scenario in which we get woken up by the OS for a background app refresh and do not want to generate any pushes).

**Comment by marcoconti83** [comment]
In theory though, if we get woken up by background app refresh (you mean background fetch, like every morning at 7am, right?), and there are events in the notification stream that we did not "see" yet, it means that the user never received a push for them, so the user is missing notifications. Now that we always fetch by notification stream, shouldn't we always create local notifications when we are in the background, as we never receive that data in any other way (won't be in the APNS payload anymore)? Or am I missing one case?

**Comment by daehn** [comment]
That's correct, the only issue there is that the user would have two notifications (one fallback and the full text fetched through the notification stream) in that case, which might be undesired.

**Comment by marcoconti83** [comment]
Thanks for clarifying
